waldo/data/profile.go

```diff
@@ -0,0 +1,197 @@
+package data
+
+import (
+	"errors"
+	"os"
+	"path/filepath"
+
+	"github.com/waldoapp/waldo-go-cli/lib"
+	"github.com/waldoapp/waldo-go-cli/lib/tpw"
+)
+
+const (
+	prfFormatVersion = 1
+)
+
+//-----------------------------------------------------------------------------
+
+type Profile struct {
+	FormatVersion int    `yaml:"format_version"`
+	UserToken     string `yaml:"user_token,omitempty"`
+
+	basePath    string // absolute
+	dirty       bool
+	profilePath string // absolute
+}
+
+//-----------------------------------------------------------------------------
+
+func SetupProfile(ck CreateKind) (*Profile, bool, error) {
+	var (
+		dataPath string
+		create   bool
+		err      error
+	)
+
+	switch ck {
+	case CreateKindAlways:
+		dataPath, err = makeHomeDataPath()
+
+		if err != nil {
+			return nil, false, err
+		}
+
+		create = true
+
+	case CreateKindIfNeeded:
+		dataPath = findHomeDataPath()
+
+		if len(dataPath) == 0 {
+			dataPath, err = makeHomeDataPath()
+
+			if err != nil {
+				return nil, false, err
+			}
+
+			create = true
+		}
+
+	default: // incl. CreateKindNever
+		dataPath = findHomeDataPath()
+
+		if len(dataPath) == 0 {
+			return nil, false, errors.New("Waldo profile not found")
+		}
+	}
+
+	prf := &Profile{
+		basePath:    filepath.Dir(dataPath),
+		profilePath: filepath.Join(dataPath, "profile.yml")}
+
+	if lib.IsRegularFile(prf.profilePath) {
+		if err := prf.load(); err != nil {
+			return nil, false, err
+		}
+
+		if err := prf.migrate(); err != nil {
+			return nil, false, err
+		}
+
+		if err := prf.Save(); err != nil {
+			return nil, false, err
+		}
+
+		return prf, false, nil
+	}
+
+	if create {
+		if err := os.MkdirAll(dataPath, 0755); err != nil {
+			return nil, false, err
+		}
+
+		prf.FormatVersion = prfFormatVersion
+
+		prf.MarkDirty()
+
+		if err := prf.Save(); err != nil {
+			return nil, false, err
+		}
+
+		return prf, true, nil
+	}
+
+	return nil, false, errors.New("Waldo profile not found")
+}
+
+//-----------------------------------------------------------------------------
+
+func (prf *Profile) BasePath() string {
+	return prf.basePath
+}
+
+func (prf *Profile) IsDirty() bool {
+	return prf.dirty
+}
+
+func (prf *Profile) MarkDirty() {
+	prf.dirty = true
+}
+
+func (prf *Profile) Path() string {
+	return prf.profilePath
+}
+
+func (prf *Profile) Save() error {
+	if !prf.IsDirty() {
+		return nil
+	}
+
+	data, err := tpw.EncodeToYAML(prf)
+
+	if err != nil {
+		return err
+	}
+
+	if err := os.WriteFile(prf.profilePath, data, 0644); err != nil {
+		return err
+	}
+
+	prf.dirty = false
+
+	return nil
+}
+
+//-----------------------------------------------------------------------------
+
+func findHomeDataPath() string {
+	dirPath, err := os.UserHomeDir()
```

Comment:
Application specific data should be placed in [UserConfigDir](https://pkg.go.dev/os#UserConfigDir) or another OS-specified not directly in the user's home dir.

Reply:
Not in the new scheme of things. Follow the link above.

Reply:
The profile is not application-specific. It is very much tied to a user. There are other changes coming that will effect the application-specific recipes.

Reply:
`UserConfigDir` is the operating-system specified directory for applications to place their user-specific configuration files.

Placing things in the user's home hasn't been recommended for some time on linux and has never been on windows or macos afaik

* https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html#variables
* https://developer.apple.com/library/archive/documentation/General/Conceptual/MOSXAppProgrammingGuide/AppRuntime/AppRuntime.html#//apple_ref/doc/uid/TP40010543-CH2-SW1 ("A few directories for mac apps")

Reply:
While I agree that technically the profile _should_ be placed in `~/Library/Application Support` (on macOS), in reality many, many apps/tools/etc. still use `~` (for example, Appium, AWS, Docker, BrowserStack, Cargo, CocoaPods, Fastlane, git, Gradle, npm, nvm, SauceLabs, Rust, SwiftPM, yarn, … Shall I go on?)

Reply:
Ok I know, i'm just very sad that we're adding to this situation :(

Reply:
BTW SwiftPM has migrated and only let the old file for old xcode (I was surprised that they were under Apple umbrella and not respecting the OS specs

https://github.com/apple/swift-package-manager/blob/da66bb12fcdb5b7c2f28b6165ded73e69fbbe5b1/Sources/Basics/FileSystem/FileSystem%2BExtensions.swift#L362

Reply:
Well now! That makes _all_ the difference. 😉